### PR TITLE
Use in-memory container registry to build plugin packages using `builder` plugin

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

This change removes the requirement of passing an external container registry to build plugin packages by using an implementation of a local registry server within the code and uses that to build plugin packages.

As part of https://github.com/vmware-tanzu/tanzu-cli/pull/370 where we removed the dependency of the builder plugin on docker and imgpkg CLI, we had to add an additional required parameter `--oci-registry` to build plugin packages.

With this change, we are still keeping that parameter but making it optional. This means if the user does not provide any external container registry, the builder plugin will spin up a local in-memory registry and will use that registry to build plugin packages.

### Which issue(s) this PR fixes

Related to #372

### Describe testing done for PR

```
$ make prepare-builder
cd cmd/plugin/builder && go build -o /Users/anujc/Documents/code/tanzu-cli/bin/builder .
```

Use updated builder plugin to build local builder and test plugin packages and publish them to localhost:5001
```
$ make plugin-publish-packages
/Users/anujc/Documents/code/tanzu-cli/bin/builder plugin build-package \
                --binary-artifacts /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins \
                --package-artifacts /Users/anujc/Documents/code/tanzu-cli/artifacts/packages
2023-07-20T14:54:41-07:00 [i] starting local registry server on port 58595, logs available at /var/folders/xc/12mmbpb50qxfb3ch8vv9nvwm0000gq/T/994070398
2023-07-20T14:54:41-07:00 [i] Using plugin binary artifacts from "/Users/anujc/Documents/code/tanzu-cli/artifacts/plugins"
2023-07-20T14:54:41-07:00 [i] 🐯 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/windows/amd64/global/test/v1.0.0-dev/tanzu-test-windows_amd64.exe
2023-07-20T14:54:41-07:00 [i] 🐵 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/linux/amd64/global/builder/v1.0.0-dev/tanzu-builder-linux_amd64
2023-07-20T14:54:41-07:00 [i] 🐼 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/darwin/amd64/global/builder/v1.0.0-dev/tanzu-builder-darwin_amd64
2023-07-20T14:54:41-07:00 [i] 🐶 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/windows/amd64/global/builder/v1.0.0-dev/tanzu-builder-windows_amd64.exe
2023-07-20T14:54:41-07:00 [i] 🐱 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/linux/amd64/global/test/v1.0.0-dev/tanzu-test-linux_amd64
2023-07-20T14:54:41-07:00 [i] 🦁 Validating Plugin Binary file: /Users/anujc/Documents/code/tanzu-cli/artifacts/plugins/darwin/amd64/global/test/v1.0.0-dev/tanzu-test-darwin_amd64
2023-07-20T14:54:41-07:00 [i] 🐯 Generating plugin package for 'plugin:test' 'target:global' 'os:windows' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:41-07:00 [i] 🦁 Generating plugin package for 'plugin:test' 'target:global' 'os:darwin' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:41-07:00 [i] 🐱 Generating plugin package for 'plugin:test' 'target:global' 'os:linux' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:41-07:00 [i] 🐵 Generating plugin package for 'plugin:builder' 'target:global' 'os:linux' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:41-07:00 [i] 🐼 Generating plugin package for 'plugin:builder' 'target:global' 'os:darwin' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:41-07:00 [i] 🐶 Generating plugin package for 'plugin:builder' 'target:global' 'os:windows' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:43-07:00 [i] 🐶 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/windows/amd64/global/builder/v1.0.0-dev/builder-windows_amd64.tar"
2023-07-20T14:54:45-07:00 [i] 🐵 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/linux/amd64/global/builder/v1.0.0-dev/builder-linux_amd64.tar"
2023-07-20T14:54:48-07:00 [i] 🐼 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/darwin/amd64/global/builder/v1.0.0-dev/builder-darwin_amd64.tar"
2023-07-20T14:54:48-07:00 [i] 🐱 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/linux/amd64/global/test/v1.0.0-dev/test-linux_amd64.tar"
2023-07-20T14:54:48-07:00 [i] 🐯 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/windows/amd64/global/test/v1.0.0-dev/test-windows_amd64.tar"
2023-07-20T14:54:48-07:00 [i] 🦁 Generated plugin package at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/darwin/amd64/global/test/v1.0.0-dev/test-darwin_amd64.tar"
2023-07-20T14:54:48-07:00 [i] Saved plugin manifest at "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages/plugin_manifest.yaml"
2023-07-20T14:54:48-07:00 [i] shutting down local registry server...

====================

docker stop temp-package-registry && docker rm -v temp-package-registry || true
temp-package-registry
temp-package-registry
docker run -d -p 5001:5000 --name temp-package-registry mirror.gcr.io/library/registry:2.7.1
5a8d5d5fe1af81060721ca34d4d3003e896220a5b2c00af42dfdbc04281fc7b7

====================

/Users/anujc/Documents/code/tanzu-cli/bin/builder plugin publish-package \
                --package-artifacts /Users/anujc/Documents/code/tanzu-cli/artifacts/packages \
                --publisher tzcli \
                --vendor vmware \
                --repository localhost:5001/test/v1/tanzu-cli/plugins
2023-07-20T14:54:50-07:00 [i] using plugin package artifacts from "/Users/anujc/Documents/code/tanzu-cli/artifacts/packages"
2023-07-20T14:54:50-07:00 [i] 🐼 publishing plugin 'name:builder' 'target:global' 'os:darwin' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:50-07:00 [i] 🐯 publishing plugin 'name:test' 'target:global' 'os:windows' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:50-07:00 [i] 🐱 publishing plugin 'name:test' 'target:global' 'os:linux' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:50-07:00 [i] 🦁 publishing plugin 'name:test' 'target:global' 'os:darwin' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:50-07:00 [i] 🐶 publishing plugin 'name:builder' 'target:global' 'os:windows' 'arch:amd64' 'version:v1.0.0-dev'
2023-07-20T14:54:50-07:00 [i] 🐵 publishing plugin 'name:builder' 'target:global' 'os:linux' 'arch:amd64' 'version:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/darwin/amd64/global/builder@sha256:4e368a18657b554d3862b71d3b9afe670771d04717909f3baa519c043a52ee42
2023-07-20T14:54:53-07:00 [i] 🐼 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/darwin/amd64/global/builder:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/windows/amd64/global/builder@sha256:3ff0d4085490bf6ea5c9405c675a9ef1509f6f25ca6da7c65d937ba461736042
2023-07-20T14:54:54-07:00 [i] 🐶 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/windows/amd64/global/builder:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/linux/amd64/global/builder@sha256:7c8044c0c519f8c6241dcc00df860d892bfb23589b02ecc8b6742cac93cc4b3f
2023-07-20T14:54:54-07:00 [i] 🐵 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/linux/amd64/global/builder:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/linux/amd64/global/test@sha256:94ba60d9d897f77f1455d45564343e2ae8bf3c8ec1a8239e9246875e01a0d554
2023-07-20T14:54:54-07:00 [i] 🐱 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/linux/amd64/global/test:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/windows/amd64/global/test@sha256:0aac66b51b033e65339b0038850c0e8961386e224352ee5dde65452ccdce8847
2023-07-20T14:54:55-07:00 [i] 🐯 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/windows/amd64/global/test:v1.0.0-dev'
localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/darwin/amd64/global/test@sha256:96ed195d0c3914d050dfdd56003f986808e497ba6bd7d7b625aa96a6a9c22071
2023-07-20T14:54:55-07:00 [i] 🦁 published plugin at 'localhost:5001/test/v1/tanzu-cli/plugins/vmware/tzcli/darwin/amd64/global/test:v1.0.0-dev'
~/D/c/tanzu-cli (use-crane-registry-serve|✔) $ 
```

### Release note

```release-note
None
```

### CODE REVIEW

File: cmd/plugin/builder/crane/interface.go [hunk 1]
```diff
@@ -10,6 +10,9 @@ type CraneWrapper interface {
 	SaveImage(image, pluginTarFilePath string) error
 	// PushImage publish the tar file to remote container registry
 	PushImage(pluginTarFilePath, image string) error
+	// ServeRegistry start an in-memory registry server
+	// returns the port on which server is running, function to shutdown the server, error if any
+	ServeRegistry(port string) (string, func(), error)
```
Comment (marckhouzam) [inline cmd/plugin/builder/crane/interface.go:15]: What is the `port` argument used for?   Is it a recommended port value?

**Review by marckhouzam — COMMENTED** [review]
This is great.  Much nicer for the plugin dev!

One comment is that the new `ServeRegistry()` function does not seem to use any logic from crane.  My guess is that you started wanting to use crane, but then found that you could do it with the http package, which is better.  But then I feel the new function doesn't really belong in the crane interface...  Maybe it should be in the `helpers` package instead?

**Comment by anujc25** [comment]
> One comment is that the new ServeRegistry() function does not seem to use any logic from crane. My guess is that you started wanting to use crane, but then found that you could do it with the http package, which is better. But then I feel the new function doesn't really belong in the crane interface... Maybe it should be in the helpers package instead?

Good point. Moved the function `ServeLocalRegistry` under `pkg/registry/helpers.go`.

File: cmd/plugin/builder/crane/interface.go (hunk 1 above)
Comment (anujc25) [reply cmd/plugin/builder/crane/interface.go]: It is an optional argument if the user wants to run the server on any specific port.

The previous comment was not explaining it that clearly so I have updated the godoc comment for the function.

> // ServeLocalRegistry start an in-memory localhost registry server at provided port
> // If the port is not provided it uses a random available port to start the registry server
> // It returns the port on which the server is running, the function to shutdown the server, error if any
> func ServeLocalRegistry(port string) (string, func(), error) {

**Review by marckhouzam — APPROVED** [review]
LGTM thanks!

**Pull request merged**